From bc1b8795ee5a0a23e9313d1b6aa639211c5d67ad Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 2 Feb 2024 10:15:24 +0100
Subject: [PATCH 0728/1695] drivers: crypto: stm32_saes: fallback to software
 on 192bit AES keys

Implement AES software operation for 192 bits keys as these are not
supported by the STM32 SAES peripheral. For that purpose ciphering final,
context copy and context freeing operations common functions are split
into CRYP/SAES peripheral specific functions.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ddf5085bbbe18a3fec9610656b4e12768d9e9dda drivers: crypto: stm32_saes: fallback to software on 192bit AES keys]
Reviewed-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Change-Id: Iab63801c9bc53bf873f77d6b1ffaa12d9ff9f9f6
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/372866
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/crypto/stm32/cipher.c | 158 ++++++++++++++++++++++++-----
 1 file changed, 135 insertions(+), 23 deletions(-)

diff --git a/core/drivers/crypto/stm32/cipher.c b/core/drivers/crypto/stm32/cipher.c
index cc95e8c8b..2cc1a6ebf 100644
--- a/core/drivers/crypto/stm32/cipher.c
+++ b/core/drivers/crypto/stm32/cipher.c
@@ -27,6 +27,9 @@ struct cryp_ctx {
 struct saes_ctx {
 	struct stm32_saes_context ctx;
 	enum stm32_saes_chaining_mode algo;
+	/* Fallback to software implementation on 192bit AES key */
+	bool use_fallback;
+	struct crypto_cipher_ctx *fallback_ctx;
 };
 
 /*
@@ -47,6 +50,8 @@ struct ip_cipher_ops {
 			   const uint8_t *iv, size_t iv_len);
 	TEE_Result (*update)(union ip_ctx *ctx, bool last_block, uint8_t *src,
 			     uint8_t *dst, size_t len);
+	void (*final)(union ip_ctx *ctx);
+	void (*copy_state)(union ip_ctx *dst_ctx, union ip_ctx *src_ctx);
 };
 
 struct stm32_cipher_ctx {
@@ -87,9 +92,17 @@ static TEE_Result cryp_update(union ip_ctx *ip_ctx, bool last_block,
 	return stm32_cryp_update(&ip_ctx->cryp.ctx, last_block, src, dst, len);
 }
 
+static void cryp_copy_state(union ip_ctx *dst_ip_ctx, union ip_ctx *src_ip_ctx)
+{
+	assert(IS_ENABLED(CFG_STM32_CRYP));
+
+	memcpy(&dst_ip_ctx->cryp, &src_ip_ctx->cryp, sizeof(dst_ip_ctx->cryp));
+}
+
 static const struct ip_cipher_ops cryp_ops = {
 	.init = cryp_init,
 	.update = cryp_update,
+	.copy_state = cryp_copy_state,
 };
 
 static TEE_Result saes_init(union ip_ctx *ip_ctx, bool is_decrypt,
@@ -101,6 +114,34 @@ static TEE_Result saes_init(union ip_ctx *ip_ctx, bool is_decrypt,
 	if (!IS_ENABLED(CFG_STM32_SAES))
 		return TEE_ERROR_NOT_IMPLEMENTED;
 
+	if (key_len == AES_KEYSIZE_192) {
+		struct crypto_cipher_ctx *ctx = ip_ctx->saes.fallback_ctx;
+		TEE_OperationMode mode = TEE_MODE_ILLEGAL_VALUE;
+		TEE_Result res = TEE_ERROR_GENERIC;
+
+		if (!IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK)) {
+			EMSG("STM32 SAES does not support 192bit keys");
+
+			return TEE_ERROR_NOT_IMPLEMENTED;
+		}
+
+		if (is_decrypt)
+			mode = TEE_MODE_DECRYPT;
+		else
+			mode = TEE_MODE_ENCRYPT;
+
+		res = ctx->ops->init(ctx, mode, key, key_len, NULL, 0, iv,
+				     iv_len);
+		if (res)
+			return res;
+
+		ip_ctx->saes.use_fallback = true;
+
+		return TEE_SUCCESS;
+	}
+
+	ip_ctx->saes.use_fallback = false;
+
 	return stm32_saes_init(&ip_ctx->saes.ctx, is_decrypt, ip_ctx->saes.algo,
 			       key_sel, key, key_len, iv, iv_len);
 }
@@ -111,12 +152,52 @@ static TEE_Result saes_update(union ip_ctx *ip_ctx, bool last_block,
 	if (!IS_ENABLED(CFG_STM32_SAES))
 		return TEE_ERROR_NOT_IMPLEMENTED;
 
+	if (ip_ctx->saes.use_fallback) {
+		struct crypto_cipher_ctx *ctx = ip_ctx->saes.fallback_ctx;
+
+		assert(IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK));
+
+		return ctx->ops->update(ctx, last_block, src, len, dst);
+	}
+
 	return stm32_saes_update(&ip_ctx->saes.ctx, last_block, src, dst, len);
 }
 
+static void saes_final(union ip_ctx *ip_ctx)
+{
+	struct crypto_cipher_ctx *ctx = ip_ctx->saes.fallback_ctx;
+
+	assert(IS_ENABLED(CFG_STM32_SAES));
+
+	if (ip_ctx->saes.use_fallback) {
+		assert(IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK));
+		ctx->ops->final(ctx);
+	}
+}
+
+static void saes_copy_state(union ip_ctx *dst_ip_ctx, union ip_ctx *src_ip_ctx)
+{
+	struct saes_ctx *src_ctx = &src_ip_ctx->saes;
+	struct crypto_cipher_ctx *fb_ctx = src_ctx->fallback_ctx;
+
+	assert(IS_ENABLED(CFG_STM32_SAES));
+
+	memcpy(&dst_ip_ctx->saes.ctx, &src_ctx->ctx, sizeof(src_ctx->ctx));
+
+	dst_ip_ctx->saes.algo = src_ctx->algo;
+	dst_ip_ctx->saes.use_fallback = src_ctx->use_fallback;
+
+	if (src_ctx->use_fallback) {
+		assert(IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK));
+		fb_ctx->ops->copy_state(dst_ip_ctx->saes.fallback_ctx, fb_ctx);
+	}
+}
+
 static const struct ip_cipher_ops saes_ops = {
 	.init = saes_init,
 	.update = saes_update,
+	.final = saes_final,
+	.copy_state = saes_copy_state,
 };
 
 static struct stm32_cipher_ctx *
@@ -147,14 +228,18 @@ static TEE_Result stm32_cipher_update(struct drvcrypt_cipher_update *dupdate)
 
 static void stm32_cipher_final(void *ctx __unused)
 {
+	struct stm32_cipher_ctx *c = to_stm32_cipher_ctx(ctx);
+
+	if (c->ops->final)
+		c->ops->final(&c->ip_ctx);
 }
 
 static void stm32_cipher_copy_state(void *dst_ctx, void *src_ctx)
 {
-	struct stm32_cipher_ctx *src = to_stm32_cipher_ctx(src_ctx);
-	struct stm32_cipher_ctx *dst = to_stm32_cipher_ctx(dst_ctx);
+	struct stm32_cipher_ctx *src_c = to_stm32_cipher_ctx(src_ctx);
+	struct stm32_cipher_ctx *dst_c = to_stm32_cipher_ctx(dst_ctx);
 
-	memcpy(dst, src, sizeof(*dst));
+	src_c->ops->copy_state(&dst_c->ip_ctx, &src_c->ip_ctx);
 }
 
 static TEE_Result alloc_cryp_ctx(void **ctx, enum stm32_cryp_algo_mode algo)
@@ -197,48 +282,75 @@ static TEE_Result stm32_cryp_cipher_allocate(void **ctx, uint32_t algo)
 	}
 }
 
-static TEE_Result alloc_saes_ctx(void **ctx, enum stm32_saes_chaining_mode algo)
+static void stm32_cryp_cipher_free(void *ctx)
 {
-	struct stm32_cipher_ctx *c = calloc(1, sizeof(*c));
-
-	if (!c)
-		return TEE_ERROR_OUT_OF_MEMORY;
-
-	FMSG("Using SAES %d", algo);
-	c->ip_ctx.saes.algo = algo;
-	c->ops = &saes_ops;
-	*ctx = &c->c_ctx;
+	struct stm32_cipher_ctx *c = to_stm32_cipher_ctx(ctx);
 
-	return TEE_SUCCESS;
+	free(c);
 }
 
 static TEE_Result stm32_saes_cipher_allocate(void **ctx, uint32_t algo)
 {
-	/*
-	 * Convert TEE_ALGO id to internal id
-	 */
+	enum stm32_saes_chaining_mode saes_algo = STM32_SAES_MODE_ECB;
+	struct crypto_cipher_ctx *fallback_ctx = NULL;
+	struct stm32_cipher_ctx *saes_ctx = NULL;
+	TEE_Result res = TEE_SUCCESS;
+
 	switch (algo) {
 	case TEE_ALG_AES_ECB_NOPAD:
-		return alloc_saes_ctx(ctx, STM32_SAES_MODE_ECB);
+		saes_algo = STM32_SAES_MODE_ECB;
+		if (IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK))
+			res = crypto_aes_ecb_alloc_ctx(&fallback_ctx);
+		break;
 	case TEE_ALG_AES_CBC_NOPAD:
-		return alloc_saes_ctx(ctx, STM32_SAES_MODE_CBC);
+		saes_algo = STM32_SAES_MODE_CBC;
+		if (IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK))
+			res = crypto_aes_cbc_alloc_ctx(&fallback_ctx);
+		break;
 	case TEE_ALG_AES_CTR:
-		return alloc_saes_ctx(ctx, STM32_SAES_MODE_CTR);
+		saes_algo = STM32_SAES_MODE_CTR;
+		if (IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK))
+			res = crypto_aes_ctr_alloc_ctx(&fallback_ctx);
+		break;
 	default:
 		return TEE_ERROR_NOT_IMPLEMENTED;
 	}
+	if (res)
+		return res;
+
+	saes_ctx = calloc(1, sizeof(*saes_ctx));
+	if (!saes_ctx) {
+		if (IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK))
+			fallback_ctx->ops->free_ctx(fallback_ctx);
+
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	FMSG("Using SAES %d", saes_algo);
+	saes_ctx->ip_ctx.saes.algo = saes_algo;
+	saes_ctx->ops = &saes_ops;
+	saes_ctx->ip_ctx.saes.fallback_ctx = fallback_ctx;
+	*ctx = &saes_ctx->c_ctx;
+
+	return TEE_SUCCESS;
 }
 
-static void stm32_cipher_free(void *ctx)
+static void stm32_saes_cipher_free(void *ctx)
 {
 	struct stm32_cipher_ctx *c = to_stm32_cipher_ctx(ctx);
 
+	if (IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK)) {
+		struct crypto_cipher_ctx *fb_ctx = c->ip_ctx.saes.fallback_ctx;
+
+		fb_ctx->ops->free_ctx(fb_ctx);
+	}
+
 	free(c);
 }
 
 static struct drvcrypt_cipher driver_cipher_cryp = {
 	.alloc_ctx = stm32_cryp_cipher_allocate,
-	.free_ctx = stm32_cipher_free,
+	.free_ctx = stm32_cryp_cipher_free,
 	.init = stm32_cipher_initialize,
 	.update = stm32_cipher_update,
 	.final = stm32_cipher_final,
@@ -247,7 +359,7 @@ static struct drvcrypt_cipher driver_cipher_cryp = {
 
 static struct drvcrypt_cipher driver_cipher_saes = {
 	.alloc_ctx = stm32_saes_cipher_allocate,
-	.free_ctx = stm32_cipher_free,
+	.free_ctx = stm32_saes_cipher_free,
 	.init = stm32_cipher_initialize,
 	.update = stm32_cipher_update,
 	.final = stm32_cipher_final,
-- 
2.47.2

